src/easier68k/core/models/list_file.py

```diff
@@ -0,0 +1,157 @@
+import json
+import re
+
+"""
+List File
+
+Represents the output from the assembler that contains all of the instructions and where in the
+destination memory they should end up.
+"""
+MAX_MEMORY_LOCATION = 16777216  # 2^24
+
+class ListFile:
+    """
+    Represents assembled instructions and their locations in memory
+    """
+    def __init__(self):
+        """
+        Constructor
+        """
+        # the keys of data must be strings so that they will work with JSON
+        # but when working with it, integers are a lot cleaner and make more sense
+        # so all of the interfaces that work with it are going to use ints
+        # but internally it will use strings
+        self.data = {}
+        self.symbols = {}
+
+    def insert_data(self, location: int, data: str):
+        """
+        Inserts the data at the given location into the list file
+        This data should be a string of hexadecimal data
+        :param location:
+        :param data:
+        :return:
+        """
+        assert location >= 0, 'Location is out of bounds!'
+        assert location < MAX_MEMORY_LOCATION, 'Location is out of bounds!'
+
+        # ensure that the data is valid
+        # loop through every 2 or 1 characters
+        for s in re.findall('..?', data):
+            # try to convert to an int of base 16
+            # just to ensure that it is valid
+            int(data, 16)
+
+        self.data[str(location)] = data
+
+    def insert_data_at_symbol(self, name: str, data: str):
+        """
+        Inserts the data at the location for the given symbol
+        :param name:
+        :param data:
+        :return:
+        """
+        self.insert_data(self.get_symbol_location(name), data)
+
+    def clear_location(self, location: int):
+        """
+        Clears the data at the given location
+        :param location:
+        :return:
+        """
+        assert location >= 0, 'Location is out of bounds!'
+        assert location < MAX_MEMORY_LOCATION, 'Location is out of bounds!'
+        assert str(location) in self.data, 'Location not defined in data!'
+
+        self.data.pop(str(location), None)
+
+    def define_symbol(self, name: str, location: int):
+        """
+        Defines a label and it's associated location
+        :param name:
+        :param location:
+        :return:
+        """
+        assert location >= 0, 'Location is out of bounds!'
+        assert location < MAX_MEMORY_LOCATION, 'Location is out of bounds!'
+
+        # check that the symbol name is a single word
+        assert re.match(r'\A[\w-]+\Z', name), 'Symbol name was not a single word!'

```

Comment:
Just checking, does this match numbers as well? Labels can contain numbers

Reply:
not certain, I'll add more tests for that